hugo/content/docs/ast-types.md

```diff
@@ -3,4 +3,340 @@ title: "AST Type Inference"
+{{</mermaid>}}
+We can see that the nested `right -> right` property in the tree is counter-intuitive and we would like to remove one level of nesting from the tree. 
+
+This can be done by reworking the grammar and adding an assigned action:
+```
+Definition: 
+    'def' name=ID ':' expr=Addition ';';
+
+Expression:
+    '(' Addition ')' | value=NUMBER;
+
+Addition infers Expression:
+    Expression ({infer Addition.left=current} '+' right=Expression)*;
+```
+Parsing the same document now leads to this AST:
+{{<mermaid>}}
+graph TD;
+expr((expr)) --> left((left))
+expr --> right((right))
+left --> left_left((left))
+left --> left_right((right))
+right --> right_v{3}
+left_left --> left_left_v{1}
+left_right --> left_right_{2}
+{{</mermaid>}}
+
+## Declared Types
+Because type inference takes into account every entity of a parser rule, even the smallest changes can update the inferred types. This can lead to an unexpected type system and incorrect behavior of services that depend on it. To minimize the risk of introducing breaking changes when modifying the grammar, we recommend to use *declared types*. This is especially true for more mature and complex languages, where a stable type system is key and breaking changes introduced by inferred types can be hard to detect. Declared types allow the user to **fix** the type of the parser rules and rely on the power of validation errors to detect breaking changes.
+
+Let's look at the example from the previous section:
+```
+X infers MyType: name=ID;
+Y infers MyType: name=ID count=INT;
+
+// should be replaced by:
+interface MyType {
+    name: string
+    count?: number
+}
+
+X returns MyType: name=ID;
+Y returns MyType: name=ID count=INT;
+```
+We now explicitly declare `MyType` directly in the grammar with the keyword `interface`. The parser rules `X` and `Y` creating AST nodes of type `MyType` need to explicitly declare the type of the AST node they create with the keyword `returns`.
+
+Contrary to [inferred types](#inferred-types), all properties must be explicitly declared in order to be valid inside of a parser rule. The following syntax:
+```
+Z returns MyType: name=ID age=INT;
+```
+will throw the following validation error `A property 'age' is not expected` because the declaration of `MyType` does not include the property `age`. In short, *declared types* add a layer of safety via validation to the grammar that prevents mismatches between the generated AST types and what the user expects.
+
+A declared type can also extend another declared type:
+```
+interface MyType {
+    name: string
+}
+
+interface MyOtherType extends MyType {
+    count: number
+}
+
+Y returns MyOtherType: name=ID count=INT;
+```
+
+Explicitly declaring type aliases in the grammar is achieved with the keyword `type`:
+```
+type X = A | B;
+
+// generates:
+type X = A | B;
+```
+
+Please note that it is not allowed to use an alias type as a return type in a parser rule. The following syntax is invalid:
```

Comment:
```suggestion
Please note that you cannot use union types as a return type of a parser rule. The following is invalid:
```

Reply:
Seems like this is actually not enforced anymore. I had a check this morning and it was not returning a validation error.

Reply:
Well, yes and no. Ideally, it should work like this:

```
type X = A | B;

ARule returns A: ...;
BRule returns B: ...; 
XRule returns X: ARule | BRule; // Valid

YRule returns X: name=ID; // Invalid
```

But you're right, it's anyway incorrect in my suggestion. We'll have to think about this a bit. Do you mind putting this part in a comment as long as we don't have a decision on it?